explorer/ast/pattern.h

```diff
@@ -299,7 +299,13 @@ class GenericBinding : public Pattern {
 
   // The index of this binding, which is the number of bindings that are in
   // scope at the point where this binding is declared.
-  auto index() const -> int { return *index_; }
+  auto index() const -> int {
+    if (index_) {
+      return *index_;
+    } else {
+      return 0;
+    }
+  }
```

Comment:
If possible, I'd like to assert in `Bindings` that we never add a value for a `GenericBinding` whose index hasn't been set yet. The index should be set by type-checking, and I think we shouldn't ever have a value for a generic binding if we don't yet know its type.

Reply:
Done, 
Inside `TypeChecker::DeclareImplDeclaration`, for adding deduced parameters to the scope ` generic_bindings.push_back(deduced)` was being directly used, I replaced that with `CollectAndNumberGenericBindingsInPattern(...)`. Seems like it fixed that issues.